src/test/java/online/carsharing/service/impl/RentalServiceImplTest.java

```diff
@@ -0,0 +1,341 @@
+    }
+
+    @Test
+    @DisplayName("Save rental throws EntityNotFoundException for non-existent car")
+    public void save_NonExistentCar_ThrowsEntityNotFoundException() {
+        RentalRequestDto rentalRequestDto = getRentalCreateDto(validRental);
+
+        when(carRepository.findById(CAR_ID)).thenReturn(Optional.empty());
+
+        assertThrows(EntityNotFoundException.class,
+                () -> rentalService.save(customer, rentalRequestDto));
+    }
+
+    @Test
+    @DisplayName("Save rental throws EntityNotFoundException for invalid user ID")
+    public void save_InvalidUserId_ThrowsEntityNotFoundException() {
+        RentalRequestDto rentalDto = getRentalCreateDto(validRental);
+
+        when(carRepository.findById(rentalDto.getCarId())).thenReturn(Optional.of(car));
+
+        assertThrows(EntityNotFoundException.class, () -> rentalService.save(customer, rentalDto));
+    }
+
+    @Test
+    @DisplayName("getById retrieves valid rental")
+    public void getById_ValidRental_RentalResponseDto() {
+        RentalResponseDto expectedResponseDto = getRentalResponseDto(validRental);
+
+        when(rentalRepository.findById(RENTAL_ID)).thenReturn(Optional.of(validRental));
+        when(rentalMapper.toDto(validRental)).thenReturn(expectedResponseDto);
+
+        RentalResponseDto actualRentalResponseDto = rentalService.getById(RENTAL_ID);
+
+        assertEquals(expectedResponseDto, actualRentalResponseDto);
+        verify(rentalMapper).toDto(validRental);
+    }
+
+    @Test
+    @DisplayName("getById throws EntityNotFoundException for non-existent rental")
+    public void getById_NonExistentRental_ThrowsException() {
+        when(rentalRepository.findById(NON_EXISTENT_RENTAL_ID)).thenReturn(Optional.empty());
+
+        assertThrows(EntityNotFoundException.class, () ->
+                rentalService.getById(NON_EXISTENT_RENTAL_ID));
+    }
+
+    @Test
+    @DisplayName("getUserRentals throws UnauthorizedAccessException for non-manager "
+            + "accessing another user's rentals")
+    public void getUserRentals_UnauthorizedAccess_ThrowsException() {
+        Long anotherUserId = MANAGER_ID;
+        assertThrows(UnauthorizedAccessException.class,
+                () -> rentalService.getUserRentals(customer, anotherUserId, true));
+    }
+
+    @Test
+    @DisplayName("getUserRentals retrieves active and inactive rentals of all users by manager")
+    public void getUserRentals_ManagerRetrievesAllActiveAndInactive_ListOfRentalResponseDto() {
+        List<Rental> activeRentals = generateRentals(true, ACTIVE_RENTALS_AMOUNT);
+        List<Rental> inactiveRentals = generateRentals(false, INACTIVE_RENTALS_AMOUNT);
+
+        when(rentalRepository.findAllByActive(true)).thenReturn(activeRentals);
+        when(rentalRepository.findAllByActive(false)).thenReturn(inactiveRentals);
+
+        List<RentalResponseDto> allActiveRentals =
+                rentalService.getUserRentals(manager, null, true);
```

Comment:
I don’t know if it’s a good practice to put null

Reply:
The `userId` is optional in the `getUserRentals()` method, which is why it is part of the logic